Link Previews

wojtus7 opened the pull request:
### Details
Added link preview in messages that contain links.
Up to three link previews are visible for single message but it can be changed using one variable.
If multiple the same links are provided only one preview will be displayed.

### Fixed Issues

Part of https://github.com/Expensify/App/issues/15001

Added based on:
https://docs.google.com/document/d/1317yQGmDgDnjwfoLybqNb5qVUU_jkhI2Bs3X9PQFcG4/edit#heading=h.fcvmbish8sz6

### Tests

1. Send a message containing a link or multiple links
2. See if the preview will show on your device as well as on other devices

- [ ] Verify that no errors appear in the JS console

### Offline steps

If cached earlier downloaded previews will show as usual, if there is no cache nothing will download so nothing should show up if there is no network connection.

### QA Steps

1. Send a message containing a link or multiple links
2. See if the preview will show on your device as well as on other devices

- [ ] Verify that no errors appear in the JS console

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

<img width="1649" alt="image" src="https://user-images.githubusercontent.com/42337257/231187187-6434c3d0-ce84-4913-9239-2190c6041585.png">

</details>

<details>
<summary>Mobile Web - Chrome</summary>

![IMG_8084](https://user-images.githubusercontent.com/42337257/231248219-0c8c53b6-b53d-4b10-85db-f30ccda9be13.png)
![IMG_8085](https://user-images.githubusercontent.com/42337257/231248234-bdb26aaf-3b3a-4041-aede-8e3927868a2f.png)

</details>

<details>
<summary>Mobile Web - Safari</summary>

![IMG_8082](https://user-images.githubusercontent.com/42337257/231253120-f1d68c5e-2fe1-443f-8b5d-646e90049c58.png)
![IMG_8083](https://user-images.githubusercontent.com/42337257/231253145-331c11dc-1ec2-44ec-93cb-aa2a31288638.png)


</details>

<details>
<summary>Desktop</summary>

<img width="1184" alt="image" src="https://user-images.githubusercontent.com/42337257/231194197-9de9cbd1-2f26-4fbb-8c19-908a573065ba.png">

</details>

<details>
<summary>iOS</summary>

https://user-images.githubusercontent.com/42337257/231164067-09e71bfa-f8ac-4dfc-bede-6fbf81c83af9.mov

</details>

<details>
<summary>Android</summary>

![Screenshot 2023-04-12 at 01 48 19](https://user-images.githubusercontent.com/42337257/231312401-f5467d63-2050-4096-8422-3e7fe8004b61.png)

</details>

AndrewGable commented:
Wow this is awesome! 😍

marcaaron commented on src/pages/home/report/ReportLinkPreview.js:
```diff
@@ -0,0 +1,102 @@
+    maxAmountOfPreviews: 3,
+};
+
+class ReportLinkPreview extends Component {
```
We can make this a function component.

marcaaron commented on package.json:
```diff
@@ -47,6 +47,7 @@
     "@formatjs/intl-numberformat": "^6.2.5",
     "@formatjs/intl-pluralrules": "^4.0.13",
     "@gorhom/portal": "^1.0.14",
+    "@microlink/mql": "^0.10.32",
```
Oh this is cool - I imagine/think we want to use some proxy.

wojtus7 replied:
Removed in favour of API.read() and proxy

marcaaron reviewed (commented):
Looks nice! Adding some more details about the next approach.

marcaaron commented:
cc @thienlnam

wojtus7 commented on src/pages/home/report/ReportActionItem.js:
```diff
@@ -98,7 +101,43 @@ class ReportActionItem extends Component {
+                });
+            }).catch((error) => {
+                console.error('One or more promises failed:', error);
+            });
```
Done ✅ 

Now we use proxy with API.read and I removed mql library from the app

marcaaron commented on src/libs/actions/Report.js:
```diff
@@ -524,6 +524,26 @@ function openPaymentDetailsPage(chatReportID, iouReportID) {
+            {
+                onyxMethod: CONST.ONYX.METHOD.MERGE,
+                key: `${ONYXKEYS.COLLECTION.REPORT_ACTIONS}${reportActionID}`,
+            },
```
This usually has a `value` field. e.g. if you want to have some loading state happening we could do this here or if we knew the shape of the data we expect to receive we could put that here. But in this case maybe the entire `optimisticData` can just be removed.

marcaaron commented on src/pages/home/report/LinkPreviewer.js:
```diff
@@ -0,0 +1,97 @@
+    maxAmountOfPreviews: 3,
+};
+
+const LinkPreviewer = ({linkMetadata, maxAmountOfPreviews}) => (
```
I think maybe something like this is a bit more consistent with the current style guide for how to treat the props in a function component though maybe there is a lot of inconsistency here in the codebase 😅 

```suggestion
function LinkPreviewer(props) {
    const linkMetadata = props.linkMetadata;
    const maxAmountOfPreviews = props.maxAmountOfPreviews;
    return _.map(_.take(uniqBy(linkMetadata, 'url'), maxAmountOfPreviews), (linkData) => {
```

marcaaron commented on src/pages/home/report/ReportActionItem.js:
```diff
@@ -105,6 +106,29 @@ class ReportActionItem extends Component {
     }
 
+    componentDidMount() {
+        const htmlText = this.props.action.message[0].html;
```
```suggestion
        const htmlText = lodashGet(this.props.action, ['message', 0, 'html']);
```

I'd maybe use something like `lodashGet()` here in case there is a bad message.

marcaaron replied:
**suggestion:** we can create a utility function for this like:

`ReportActionUtils.containsLink(reportAction)`

Then here just have

```js
if (!ReportActionUtils.containsLink(this.props.action)) {
    return;
}

Report.expandURLPreview(this.props.report.reportID, this.props.action.reportActionID);
```

marcaaron commented on src/pages/home/report/ReportActionItem.js:
```diff
@@ -234,6 +258,12 @@ class ReportActionItem extends Component {
                 {children}
+                {this.props.action.linkMetadata && (
+                    <LinkPreviewer
+                        maxAmountOfPreviews={3}
```
I wonder if it is needed. Maybe we should just show all the previews for now.

marcaaron reviewed (changes requested):
Looks great and thanks for the work on this! I had a few small style notes and a comment about the limiting of previews (not sure if we need to do it yet).

marcaaron commented:
@wojtus7 Please let me know if I can explain anything else or help move things along 🙇

mountiny commented on src/libs/ReportActionsUtils.js:
```diff
@@ -162,6 +162,16 @@ function getMostRecentIOURequestActionID(reportActions) {
 }
 
+/**
+ * Finds most recent IOU request action ID.
```
I believe this comment is not correct

mountiny commented on src/libs/actions/Report.js:
```diff
@@ -607,6 +607,19 @@ function openPaymentDetailsPage(chatReportID, iouReportID) {
 }
 
+/**
+ * Gets metadata info about links in provided report
```
```suggestion
 * Gets metadata info about links in the provided report
```

mountiny commented on src/pages/home/report/ReportActionItem.js:
```diff
@@ -120,6 +121,14 @@ function ReportActionItem(props) {
+        }
+
+        Report.expandURLPreview(props.report.reportID, props.action.reportActionID);
+    }, [])
```
Linter warning here

mountiny commented on src/pages/home/report/ReportActionItem.js:
```diff
@@ -279,6 +288,11 @@ function ReportActionItem(props) {
                 {children}
+                {props.action.linkMetadata && (
+                    <LinkPreviewer
+                        linkMetadata={props.action.linkMetadata}
```
I believe there should be just one link preview shown if mulitple links is in the message based on the PR body 
```suggestion
                        linkMetadata={props.action.linkMetadata}
                        maxAmountOfPreviews=1
```
<img width="956" alt="image" src="https://github.com/Expensify/App/assets/36083550/b478e039-83e7-44c1-948c-5453543ed7ca">

mountiny reviewed (changes requested):
Looking good, I will ask for some C+ for testing ~but I have noticed we should multiple of the link previews which is not intended, is it?~

I see your comment I misread that was for same links. Where do we add the limit for 3 link previews in the code?

0xmiroslav reviewed (changes requested):
@wojtus7 author checklist is outdated. please pull from latest main.

Please check and fix this everywhere.

    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.

wojtus7 commented on src/pages/home/report/LinkPreviewer.js:
```diff
@@ -0,0 +1,97 @@
+    ),
+
+    /** Maximum amount of visible link previews. */
+    maxAmountOfPreviews: PropTypes.number,
```
Added the `maxAmountOfPreviews: undefined` because without it the limit simply doesn't exist.
Or do you think I should set some value and make additional "disablePreviewsAmountLimit" boolean prop?

wojtus7 replied:
So I made it to be:

`-1` => unlimited
`0` => none are displayed

mountiny commented:
@wojtus7 have you posted the [CLA comment](https://github.com/Expensify/App/pull/17288#issuecomment-1504008555)?

wojtus7 committed:
Add Link Previews

wojtus7 committed:
Create disable links and unlimited using maxAmountOfPreviews prop

mountiny reviewed (dismissed):
Thanks @wojtus7 I see that the CLa should be done and commits are verified.

@0xmiroslav do you think you can do the checklist and testing please?

wojtus7 committed:
Replace simple check with isEmpty

wojtus7 commented on src/pages/home/report/LinkPreviewer.js:
```diff
@@ -0,0 +1,98 @@
+            <View style={styles.linkPreviewWrapper} key={url}>
+                <View style={styles.flexRow}>
+                    {logo && <Image style={styles.linkPreviewLogoImage} source={{uri: logo.url}} />}
+                    {publisher && <Text fontSize={variables.fontSizeLabel} style={styles.pl2}>{publisher}</Text>}
```
Sorry - I replaced it with `!_.isEmpty()`

wojtus7 committed:
Run prettier and add array length check

wojtus7 committed:
Use size instead of length

wojtus7 commented on src/styles/styles.js:
```diff
@@ -3257,6 +3261,27 @@ const styles = {
+        flex: 1,
+        resizeMode: 'contain',
+        maxWidth: 350,
+        maxHeight: 350,
```
Good call - I think it's a good idea to make it something like: 
`maxHeight: Math.min(350, someValueBasedOnScreenSize)`
So it will be scaled down if the screen size of the phone is too small

wojtus7 committed:
Add check for isArray

wojtus7 committed:
Run prettier

wojtus7 committed:
Show smaller preview image if image would take more than half the screen

wojtus7 commented on src/pages/home/report/LinkPreviewer.js:
```diff
@@ -0,0 +1,129 @@
+    _.map(
+        _.take(uniqBy(props.linkMetadata, 'url'), props.maxAmountOfPreviews >= 0 ? Math.min(props.maxAmountOfPreviews, props.linkMetadata.length) : props.linkMetadata.length),
+        (linkData) => {
+            if (_.isArray(linkData)) {
+                return;
+            }
```
I think so. I don't create this data myself just wait for data from backend :(

marcaaron replied:
Ah it's most likely because our stupid PHP API is turning empty objects into arrays when JSON encoding 😄  

I will make a note to fix this eventually. I think we can live with the array check for now so that this doesn't become a blocker.

wojtus7 committed:
Replace Dimensions with useWindowDimensions hook

wojtus7 committed:
Add logic to scale image width based on ratio so it doesn't move away from left side of component

mountiny reviewed (dismissed):
This looks good to me, @0xmiroslav do you have any other concerns? Would you be able to add a checklist?

mountiny commented:
great suggestions

> Feature request: Add ability to close preview (X button at the top right)

This will be a follow up, not in this PR

> Isn't it better to keep left margin when edit message?

I agree it looks better with the margin

Otherwise I agree with the performance improvements, lets try to implement all of them @wojtus7

marcaaron commented:
Just a heads up (mostly for myself and @mountiny as nobody else can see the link) we noticed there is a [server error](https://github.com/Expensify/Expensify/issues/287142) related to the API command used here. I'll give this another review now to see if there's anything obvious that could be causing it.

wojtus7 committed:
Merge remote-tracking branch 'upstream/main' into wojtus7/link-preview

wojtus7 committed:
Limit amount of ExpandURL calls, and fix styles while in draft

wojtus7 committed:
Fix resize link preview logic

wojtus7 committed:
Add max width

wojtus7 committed:
Prettier

wojtus7 committed:
Filter empty objects and arrays in linkMetadata

wojtus7 commented:
@0xmiroslav thanks for feedback!
Can you take a look? 

I think I fixed all thing except for limit to fetch only necessary urls. As far as I can see we would need to change how ExpandURLPreview APi works to achieve that?

wojtus7 committed:
Merge remote-tracking branch 'upstream/main' into wojtus7/link-preview

wojtus7 committed:
Fix styles

marcaaron commented on src/libs/ReportActionsUtils.js:
```diff
@@ -170,6 +170,28 @@ function getMostRecentIOURequestActionID(reportActions) {
+/**
+ * Returns array of links inside given report
+ *
+ * @param {reportAction} reportAction
```
```suggestion
 * @param {object} reportAction
```
there's no typedef for `reportAction` anywhere I don't think (well not yet, we would maybe have one with TS eventually)

marcaaron commented on src/libs/ReportActionsUtils.js:
```diff
@@ -170,6 +170,28 @@ function getMostRecentIOURequestActionID(reportActions) {
+    while (match !== null) {
+        urls.push(match[1]);
+        match = regex.exec(htmlContent);
+    }
```
NAB, I think maybe `string.matchAll(regex)` is a bit more obvious - but this works for sure.

marcaaron replied:
something like...

```js
return _.map([...htmlContent.matchAll(regex)], match => match[1]);
```

marcaaron commented on src/libs/ReportActionsUtils.js:
```diff
@@ -170,6 +170,28 @@ function getMostRecentIOURequestActionID(reportActions) {
+ * @param {reportAction} reportAction
+ * @returns {Boolean}
+ */
+function getLinksInsideReport(reportAction) {
```
Could we change the name to something like `extractLinksFromMessageHtml()`. We are focusing on a single report action, so let's narrow it down vs implying we're getting all links for the whole chat.

marcaaron commented on src/pages/home/report/ReportActionItem.js:
```diff
@@ -334,6 +346,11 @@ function ReportActionItem(props) {
                 {children}
+                {!_.isEmpty(props.action.linkMetadata) && (
+                    <View style={props.draftMessage ? styles.chatItemReactionsDraftRight : {}}>
+                        <LinkPreviewer linkMetadata={_.filter(props.action.linkMetadata, (item) => !_.isEmpty(item))} />
```
There are some empty items coming back? Maybe I can fix this as well on the backend.

wojtus7 replied:
That was more to avoid error mentioned here: https://github.com/Expensify/App/pull/17288#issuecomment-1568177325

There should be array of objects in LinkPreviewer types, but if there is empty `[]` instead of real object it shows a warning of wrong types. Since _.isEmpty also removes empty arrays I used it to filter. 

One can say that:
`
            if (_.isArray(linkData)) {
                return;
            }
`
 inside LinkPreviewer is redundant but I just left it just to be safe.

marcaaron reviewed (changes requested):
Sorry for the delay on the review as some people were OOO last week. 

Changes are looking good - I had a couple of comments / minor things to update.

wojtus7 committed:
Merge remote-tracking branch 'upstream/main' into wojtus7/link-preview

wojtus7 committed:
Improve extractLinksFromMessageHtml func

0xmiroslav reviewed (commented):
Some minor last changes:

wojtus7 committed:
Regex comment and add default transparency

wojtus7 commented on src/styles/styles.js:
```diff
@@ -23,7 +23,7 @@ import CONST from '../CONST';
 import userSelect from './utilities/userSelect';
 
-function getTransparentColor(color, transparency) {
+function getTransparentColor(color, transparency = 'ff') {
```
Make sense as well. Actually it says: "getTransparentColor" so I think I would personally use something like '50%' so `80` to make it half tansparent by default

wojtus7 committed:
Use empty string instead of ff for transparency

wojtus7 commented:
@0xmiroslav Hi I think it's ready. 

I think I addressed all comments except this: https://github.com/Expensify/App/pull/17288#issuecomment-1568224911 but since we show all previews right now and not limiting amount anywhere it's not relevant, and I don't know if I can do anything about it from the frontend right now.

mountiny commented on src/libs/ReportActionsUtils.js:
```diff
@@ -170,6 +170,25 @@ function getMostRecentIOURequestActionID(reportActions) {
 }
 
+/**
+ * Returns array of links inside given report
```
```suggestion
 * Returns array of links inside a given report action
```

mountiny commented on src/libs/actions/Report.js:
```diff
@@ -582,6 +582,19 @@ function readOldestAction(reportID, reportActionID) {
 }
 
+/**
+ * Gets metadata info about links in the provided report
```
I think this is linked to the report action firstly
```suggestion
 * Gets metadata info about links in the provided report action
```

mountiny commented on src/pages/home/report/LinkPreviewer.js:
```diff
@@ -0,0 +1,142 @@
+    maxAmountOfPreviews: -1,
+};
+
+const LinkPreviewer = (props) => {
```
https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#should-i-declare-my-components-with-arrow-functions-const-or-the-function-keyword

We should use function syntax for the functional components

mountiny reviewed (commented):
couple of NABs

@0xmiroslav can you please give this one another go, seems like we are close

wojtus7 committed:
Update comment

Co-authored-by: Vit Horacek <36083550+mountiny@users.noreply.github.com>

wojtus7 committed:
Update expandURLPreview comment

Co-authored-by: Vit Horacek <36083550+mountiny@users.noreply.github.com>

wojtus7 committed:
Use function keyword instead of arrow

wojtus7 committed:
Prettier

wojtus7 committed:
Don't expandURLPreview when action is delete

wojtus7 commented on src/pages/home/report/ReportActionItem.js:
```diff
@@ -117,6 +119,16 @@ function ReportActionItem(props) {
+        }
+
+        downloadedPreviews.current = urls;
+        Report.expandURLPreview(props.report.reportID, props.action.reportActionID);
```
I think I resolved it by checking if `props.action.pendingAction === CONST.RED_BRICK_ROAD_PENDING_ACTION.DELETE` before API call

marcaaron reviewed (approved):
LGTM

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

youssef-lr commented:
Possible regression, HTML URL is displayed in the LHN:

<img width="1800" alt="Screenshot 2023-06-14 at 13 20 33" src="https://github.com/Expensify/App/assets/9680864/20892dfc-e29d-4748-9c8c-78325d8798d3">

mountiny commented:
I am not sure the changes here dont really touch that. Might be still worth trying revert locally

youssef-lr commented:
Sorry folks for the scare, looks like this was fixed in the backend and I had checked out a different branch before and completely forgot.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/5272030051) to staging by https://github.com/marcaaron in version: 1.3.28-0 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|failure ❌ 
🖥 desktop 🖥|failure ❌
🍎 iOS 🍎|failure ❌ 
🕸 web 🕸|failure ❌

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/5293001429) to production by https://github.com/AndrewGable in version: 1.3.28-5 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅

iwiznia commented on src/libs/ReportActionsUtils.js:
```diff
@@ -170,6 +170,25 @@ function getMostRecentIOURequestActionID(reportActions) {
+    const regex = /<a\s+(?:[^>]*?\s+)?href="([^"]*)"/gi;
+
+    if (!htmlContent) {
+        return;
```
BTW this caused a bug which I fixed here https://github.com/Expensify/App/pull/21369/files

marcaaron replied:
Thanks for catching it 🙇

eVoloshchak commented on src/pages/home/report/ReportActionItem.js:
```diff
@@ -328,6 +340,11 @@ function ReportActionItem(props) {
                 {children}
+                {!_.isEmpty(props.action.linkMetadata) && (
+                    <View style={props.draftMessage ? styles.chatItemReactionsDraftRight : {}}>
+                        <LinkPreviewer linkMetadata={_.filter(props.action.linkMetadata, (item) => !_.isEmpty(item))} />
```
👋 This has caused a regression in https://github.com/Expensify/App/issues/21195
We shouldn't show link preview if message was flagged as inappropriate, this was resolved by hiding `LinkPreview` if `isHidden=true`

eVoloshchak commented:
👋 Coming from https://github.com/Expensify/App/issues/22888, 
Kind of an edge case, if you 'reply in thread' to a message with link preview and then delete that message, the link preview was still visible 
This was resolved by clearing `linkMetadata` in `optimisticReportActions` when [deleteReportComment](https://github.com/StevenKKC/App/blob/64e0b04e1c7497cfb9c05a5a97938cdcdfe45e7c/src/libs/actions/Report.js#L871) is called
